src/zap/zapinfo.cpp

```diff
@@ -2136,19 +2136,55 @@ DWORD FilterMethodAttribsForIsSupported(DWORD attribs, CORINFO_METHOD_HANDLE ftn
                     )
                 )
             )
+        {
             return attribs;
-#endif // defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+        }
 
-        // If it's another get_IsSupported method on some other ISA class, do not report as intrinsic
-        if (isX86intrinsic
-            || strcmp(namespaceName, "System.Runtime.Intrinsics.Arm.Arm64") == 0
-            || strcmp(namespaceName, "System.Runtime.Intrinsics") == 0)
+        // If it's an intrinsic that requires VEX encoding, do not report as intrinsic
+        // to force this to become a regular method call.
+        // We don't allow RyuJIT to use VEX encoding at AOT compilation time, so these
+        // cannot be pregenerated. Not reporting them as intrinsic will make sure
+        // it will do the right thing at runtime (the called method will be JITted).
+        // It will be slower, but correct.
+        if (fIsX86intrinsic
+            && (
+                strcmp(className, "Avx") == 0 || strcmp(className, "Fma") == 0 || strcmp(className, "Avx2") == 0 || strcmp(className, "Bmi1") == 0 || strcmp(className, "Bmi2") == 0
+                || (
+                    strcmp(className, "X64") == 0
+                    && (
+                        strcmp(enclosingClassName, "Bmi1") == 0 || strcmp(enclosingClassName, "Bmi2") == 0
+                        )
+                    )
+                )
+            )
         {
-            attribs &= ~CORINFO_FLG_JIT_INTRINSIC;
+            // We do want the IsSupported for VEX instructions to be recognized as intrinsic so that the
+            // potentially worse quality code doesn't actually run until tiered JIT starts
+            // kicking in and recompiling methods. Reporting this as intrinsic makes RyuJIT expand it
+            // into `return false`.
+            if (fIsGetIsSupportedMethod)
+                return attribs;
+
+            // Treat as a regular method call (into a JITted method).
+            return (attribs & ~CORINFO_FLG_JIT_INTRINSIC) | CORINFO_FLG_DONT_INLINE;
+        }
 
-            // The intrinsic just recursively calls itself in IL so it wouldn't inline anyway,
-            // but let's make that explicit.
-            attribs |= CORINFO_FLG_DONT_INLINE;
+#endif // defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+
+        // Do not report the get_IsSupported method as an intrinsic. This will turn the call into a regular
+        // call. We also make sure none of the hardware intrinsic method bodies get pregenerated in crossgen
+        // (see ZapInfo::CompileMethod) but get JITted instead. The JITted method will have the correct
+        // answer for the CPU the code is running on.
+        if (fIsGetIsSupportedMethod &&
```

Comment:
As per https://github.com/dotnet/coreclr/pull/24917#issuecomment-501320491, I think the ARM `IsSupported` checks should be constant on x86 as should the x86 `IsSupported` checks on ARM.

Reply:
Isn't that what this code is doing?

This logic is going to strip the intrinsic flag - so we're making sure we only strip it for x86 intrinsics on x86 and arm64 intrinsics on arm64.

Reply:
Ah, right. I misread the combination of the comment plus the if checks. 👍